Fix reference-leak issues in typecast.c, per bug #843723.


diff -Naur psycopg2-2.0.14.orig/psycopg/typecast.c psycopg2-2.0.14/psycopg/typecast.c
--- psycopg2-2.0.14.orig/psycopg/typecast.c	2010-02-28 14:29:43.000000000 -0500
+++ psycopg2-2.0.14/psycopg/typecast.c	2012-09-14 14:07:17.521810955 -0400
@@ -381,9 +381,12 @@
 {
     typecastObject *self = (typecastObject*)obj;
 
+    PyObject_GC_UnTrack(self);
+
     Py_CLEAR(self->values);
     Py_CLEAR(self->name);
     Py_CLEAR(self->pcast);
+    Py_CLEAR(self->bcast);
 
     obj->ob_type->tp_free(obj);
 }
@@ -396,6 +399,7 @@
     Py_VISIT(self->values);
     Py_VISIT(self->name);
     Py_VISIT(self->pcast);
+    Py_VISIT(self->bcast);
     return 0;
 }
 
@@ -587,7 +591,7 @@
     PyObject *old, *res = NULL;
     typecastObject *self = (typecastObject *)obj;
 
-    /* we don't incref, the caster *can't* die at this point */
+    Py_INCREF(obj);
     old = ((cursorObject*)curs)->caster;
     ((cursorObject*)curs)->caster = obj;
 
@@ -602,6 +606,7 @@
     }
 
     ((cursorObject*)curs)->caster = old;
+    Py_DECREF(obj);
 
     return res;
 }
